lib/ObjWriter/objwriter.cpp

```diff
@@ -550,22 +547,119 @@ extern "C" void EmitDebugFileInfo(ObjectWriter *OW, int FileId,
+
+  for (int i = 0; i < NumVarInfos; i++) {
+    // Emit an S_LOCAL record
+    DebugVarInfo var = LocInfos[i];
+    LocalSym sym;
+
+    memset(&sym, 0, sizeof(sym));
+
+    int sizeofSym = sizeof(LocalSym);
+    short rectyp = SymbolRecordKind::S_LOCAL;
+    short reclen = sizeof(rectyp) + sizeofSym + var.name.length() + 1;
+    TypeIndex typeIndex(var.typeIndex);
+    sym.Type = typeIndex;
 
+    if (var.isParam) {
+      sym.Flags |= LocalSym::IsParameter;
+    }
+
+    OST.EmitIntValue(reclen, 2);
+    OST.EmitIntValue(rectyp, 2);
+    OST.EmitBytes(StringRef((char *)&sym, sizeofSym));
+    OST.EmitBytes(StringRef(var.name.c_str(), var.name.length() + 1));
+
+    for (const auto &range : *var.ranges) {
+      assert(range.varNumber == var.varNumber);
+
+      // Emit a range record
+      LocalVariableAddrRange *prange = nullptr;
+
+      switch (range.loc.vlType) {
+      case ICorDebugInfo::VLT_REG:
+      case ICorDebugInfo::VLT_REG_FP: {
+        DefRangeRegisterSym rec;
+        int sizeofRec = sizeof(rec);
+        memset(&rec, 0, sizeofRec);
+        short rectyp = SymbolRecordKind::S_DEFRANGE_REGISTER;
+        short reclen = sizeofRec + sizeof(rectyp);
+        rec.Range.OffsetStart = range.startOffset;
+        rec.Range.Range = range.endOffset - range.startOffset;
+        rec.Range.ISectStart = 0;
+        rec.Register = cvRegMapAmd64[range.loc.vlReg.vlrReg];
+        prange = &rec.Range;
+        OST.EmitIntValue(reclen, 2);
+        OST.EmitIntValue(rectyp, 2);
+        OST.EmitBytes(
+            StringRef((char *)&rec, offsetof(DefRangeRegisterSym, Range)));
+        break;
+      }
+
+      case ICorDebugInfo::VLT_STK: {
+        DefRangeRegisterRelSym rec;
+        int sizeofRec = sizeof(DefRangeRegisterRelSym);
+        memset(&rec, 0, sizeofRec);
+        short rectyp = SymbolRecordKind::S_DEFRANGE_REGISTER_REL;
+        short reclen = sizeofRec + sizeof(rectyp);
+        rec.Range.OffsetStart = range.startOffset;
+        rec.Range.Range = range.endOffset - range.startOffset;
+        rec.Range.ISectStart = 0;
+        rec.BaseRegister = cvRegMapAmd64[range.loc.vlStk.vlsBaseReg];
+        rec.BasePointerOffset = range.loc.vlStk.vlsOffset;
+        prange = &rec.Range;
+        OST.EmitIntValue(reclen, 2);
+        OST.EmitIntValue(rectyp, 2);
+        OST.EmitBytes(
+            StringRef((char *)&rec, offsetof(DefRangeRegisterRelSym, Range)));
+        break;
+      }
+
+      case ICorDebugInfo::VLT_REG_BYREF:
+      case ICorDebugInfo::VLT_STK_BYREF:
+      case ICorDebugInfo::VLT_REG_REG:
+      case ICorDebugInfo::VLT_REG_STK:
+      case ICorDebugInfo::VLT_STK_REG:
+      case ICorDebugInfo::VLT_STK2:
+      case ICorDebugInfo::VLT_FPSTK:
+      case ICorDebugInfo::VLT_FIXED_VA:
+        // TODO: for optimized debugging
+        break;
+
+      default:
+        _ASSERTE(!"Unknown varloc type!");
+        break;
+      }
+
+      // Emit range
+      if (prange != nullptr) {
```

Comment:
This is about range emission which is part of symbol emission above. I like it to be a function that is invoked for each case above instead of falling thru like this.